### ISSUE TITLE
Ability to specify that T[K] extends a type

### ISSUE THREAD

**Issue by Roaders** [microsoft/TypeScript#31325]
## Search Terms

enforce key type of generic type

## Suggestion

I would like to specify that a property, defined as a function argument, of a generic type is of a specific type

## Use Cases

I want to pass an item and a key to a function and specify that the type of the property specified by the key is a certain type

## Examples

```
function cacheItem<T extends {K: string}, K extends keyof T>(item: T, key: K){
     const key = item[key]; // should be typed as string
}

const myItem = {
    id = "string",
    count = 1234
}

cacheItem(myItem, "id") // works
cacheItem(myItem, "count") // should throw error
```

## Checklist

My suggestion meets these guidelines:

* [x ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x ] This wouldn't change the runtime behavior of existing JavaScript code
* [x ] This could be implemented without emitting different JS based on the types of the expressions
* [x ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by jack-williams** [comment]
Looks related: #31275

**Comment by anion155** [comment]
```typescript
function cacheItem<
  T extends object,
  K extends keyof T,
  TK extends string = T[K] extends string ? T[K] : never,
>(item: T, key: K) {
     const value = item[key] as TK; // should is typed as string
}
```

**Comment by anion155** [comment]
```typescript
function cacheItem<
  T extends any,
  K extends {
    [K in keyof T]: T[K] extends string ? K : never
  }[keyof T]
>(item: T, key: K) {
  const value: string = item[key]; // is typed as string
  value.replace;
}
```

**Comment by jcalz** [comment]
`{K: string}` is not saying what you intend to say, since `K` is just a literal key there (e.g., `foo.K = "oops"`).  Instead, you can use a [mapped type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types) such as `{[P in K]: string}` or the [equivalent](https://github.com/Microsoft/TypeScript/blob/v3.4.5/lib/lib.es5.d.ts#L1449-L1454) `Record<K, string>`.  (Eventually you may be able to [use `K` directly as an index signature]((https://github.com/microsoft/TypeScript/pull/26797)) as in `{[k: K]: string}` but that is not currently supported.)

**Comment by RyanCavanaugh** [comment]
You don't need new machinery to do this.

```ts
function cacheItem<K extends string>(item: Record<K, string>, key: K) {
    const val = item[key]; // should be typed as string (is)
}

const myItem = {
    id: "string",
    count: 1234
}

cacheItem(myItem, "id") // works
cacheItem(myItem, "count"); // Error
```
